source/request-as-event-emitter.js

```diff
@@ -228,6 +228,9 @@ module.exports = (options = {}) => {
 				options.headers['content-length'] = uploadBodySize;
 			}
 
+			if (is.function(options.finalize)) {
+				await options.finalize(options);
+			}
```

Comment:
If `got.create` wasn't existing, then it'd be very helpful. I think this options is useless.

Reply:
I'm sorry, but we are completely misunderstanding each other.

If I understand your suggestion regarding a handler, then a custom handler would run before the default handler.

The entire point of this feature is to create an opportunity to add a signature after the request has been normalized, just before it is sent.

Here's a little test that might illustrate the difference:

```js
test('does something useful', async t => {
	const instance = got.create({
		baseUrl: `${s.url}/api/`,
		options: got.defaults.options,
		methods: got.defaults.methods,
		handler: (url, options, next) => {
			t.is(options.hostname, 'example.com')
			t.is(options.path, '/api/?foo=bar')
			return next(url, options);
		}
	});
	await instance('?foo=bar');
});
```
result:
```

  1 test failed

  does something useful

  /Users/jstewmon/dev/forks/got/test/finalize.js:95

   94:     handler: (url, options, next) => {
   95:       t.is(options.hostname, 'example.com')
   96:       t.is(options.path, '/api/?foo=bar')

  Difference:

  - undefined
  + 'example.com'

  Object.handler (test/finalize.js:95:6)
  got (source/create.js:34:20)
  test/finalize.js:100:8
```

Reply:
Thanks for clarifying this out :)

Reply:
This should be `.on('beforeRequest', ...)`. Options may be overridden.

Reply:
I considered doing that in `assignOptions`, but I decided that it was better to leave that choice to the caller, since I think it is pretty easy:

```js
const got = require('got');
const a = got.extend({beforeRequest: opts => { opts.header.foo = 'bar'; }});
const b = a.extend({beforeRequest: opts => { a.defaults.beforeRequest(opts); opts.header.bar = 'foo'; }});
```

If the functions are chained, I don't think there's a convenient way to replace the `beforeRequest` option if that's the behavior one desires.

Reply:
> we just wrap it in a function that calls both.

@sindresorhus We need to be strict here. Only for `beforeRequest` and `retries`.

> I decided that it was better to leave that choice to the caller, since I think it is pretty easy

It's OK for now. Lots of things is gonna change if #510 gets merged. #510 needs a lot of polishing.

Reply:
I took a look at #510 and had the same initial reaction as others - I don't get it. :-)

Personally, I don't see providing merge/extend features as creating a lot of leverage. Since the options are accessible on a got instance, the caller can always use a merge strategy of their choosing to create the configuration for a new client. If it were my choice, I'd probably just provide `Object.assign({}, old, new)` semantics for creating new instances.

Reply:
@jstewmon It's not as easy as it seems to be ;) That's for the future: Got plugins.